File: Find Min (Stack)/ruby-solution/stack_with_min.rb
require_relative 'min_stack'
require_relative 'node'

class StackWithMin

  attr_reader :top

  def initialize
    @top = nil
    @min = MinStack.new
  end

  def pop
    if @top
      item = @top.data
      @top = @top.next
      @min.pop
      item
    else
      puts "Empty Stack"
    end
  end

  def push(data)
    new_top = Node.new(data)
    new_top.next = @top
    @top = new_top
    @min.push(data)
  end

  def peek
    if @top
      @top.data
    else
      puts "Empty Stack"
    end
  end

  def get_min
    @min.peek
  end

end

# driver/tester code
s = StackWithMin.new
s.push(5)
p "Expected: 5. Received: #{s.get_min}"
s.push(6)
p "Expected: 5. Received: #{s.get_min}"
s.push(3)
p "Expected: 3. Received: #{s.get_min}"
s.push(3)
p "Expected: 3. Received: #{s.get_min}"
s.pop
p "Expected: 3. Received: #{s.get_min}"
s.pop
p "Expected: 5. Received: #{s.get_min}"
s.push(4)
p "Expected: 4. Received: #{s.get_min}"
s.push(7)
p "Expected: 4. Received: #{s.get_min}"

File: Find Min (Stack)/python-solution/stack_with_min.py
from minstack import MinStack
from node import Node

class StackWithMin(object):
  def __init__(self):
    self.top = None
    self.min = MinStack()

  def pop(self):
    if self.top == None:
      raise SystemExit("Empty Stack")
    item = self.top.data
    self.top = self.top.next
    self.min.pop()
    return item

  def push(self, data):
    new_top = Node(data)
    new_top.next = self.top
    self.top = new_top
    self.min.push(data)

  def peek(self):
    if self.top == None:
      raise SystemExit("Empty Stack")
    else:
      return self.top.data

  def get_min(self):
    return self.min.peek()

#tester code
s = StackWithMin()
s.push(5)
print "Expected 5:"
print s.get_min()
s.push(6)
print "Expected 5:"
print s.get_min()
s.push(3)
print "Expected 3:"
print s.get_min()
s.pop()
print "Expected 5:"
print s.get_min()
s.push(4)
print "Expected 4:"
print s.get_min()
s.push(7)
print "Expected 4:"
print s.get_min()